sockeye/layers.py

```diff
@@ -58,18 +61,376 @@ def moments(inputs: mx.sym.Symbol) -> Tuple[mx.sym.Symbol, mx.sym.Symbol]:
+    # (batch, heads, length, depth/heads)
+    x = mx.sym.transpose(data=x, axes=(0, 2, 1, 3))
+    # (batch * heads, length, depth/heads)
+    return mx.sym.reshape(data=x, shape=(-3, length, -1))
+
+
+def combine_heads(x: mx.sym.Symbol, length: int, heads: int) -> mx.sym.Symbol:
+    """
+    Returns a symbol with both batch & length, and head & depth dimensions combined.
+
+    :param x: Symbol of shape (batch * heads, length, depth_per_head).
+    :param length: Sequence length.
+    :param heads: Number of heads.
+    :return: Symbol of shape (batch * length, depth).
+    """
+    # (batch, heads, length, depth_per_head)
+    x = mx.sym.reshape(data=x, shape=(-4, -1, heads, length, 0))
+    # (batch, length, heads, depth_per_head)
+    x = mx.sym.transpose(x, axes=(0, 2, 1, 3))
+    # (batch, length, depth)
+    return mx.sym.reshape(x, shape=(-1, length, -3))
+
+
+def broadcast_lengths(x: mx.sym.Symbol, heads: int) -> mx.sym.Symbol:
+    """
+    Broadcasts the length information of each sequence to multiple heads.
+
+    :param x: Symbol(batch, 1)
+    :param heads: Number of heads.
+    :return: Symbol(batch * heads, 1)
+    """
+    # x: (batch, 1)
+    x = mx.sym.expand_dims(x, axis=1)
+    # x: (batch, heads)
+    x = mx.sym.broadcast_to(x, shape=(0, heads))
+    # x: (batch * heads, 1)
+    x = mx.sym.reshape(x, shape=(-3,))
+    return x
+
+
+def dot_attention(queries: mx.sym.Symbol,
+                  keys: mx.sym.Symbol,
+                  values: mx.sym.Symbol,
+                  length: mx.sym.Symbol,
+                  dropout: float = 0.0,
+                  bias: Optional[mx.sym.Symbol] = None):
+    """
+    Computes dot attention for a set of queries, keys, and values.
+
+    :param queries: Attention queries. Shape: (n, lq, d).
+    :param keys: Attention keys. Shape: (n, lk, d).
+    :param values: Attention values. Shape: (n, lk, dv).
+    :param length: Sequence lengths. Shape: (n,).
+    :param dropout: Dropout probability.
+    :param bias: Optional bias tensor. Shape: (1, lq, lk).
+    :return: 'Context' vectors for each query. Shape: (n, lq, dv).
+    """
+    # (n, lq, lk)
+    logits = mx.sym.batch_dot(lhs=queries, rhs=keys, transpose_b=True)
+
+    # mask lk dimension
+    # (lk, n, lq)
+    logits = mx.sym.transpose(data=logits, axes=(2, 0, 1))
+    logits = mx.sym.SequenceMask(data=logits,
+                                 use_sequence_length=True,
+                                 sequence_length=length,
+                                 value=-99999999.)
+    # (n, lq, lk)
+    logits = mx.sym.transpose(data=logits, axes=(1, 2, 0))
+
+    if bias is not None:
+        logits = mx.sym.broadcast_add(logits, bias)
+
+    probs = mx.sym.softmax(logits, axis=-1)
+    probs = mx.sym.Dropout(probs, p=dropout) if dropout > 0.0 else probs
+
+    # (n, lq, lk) x (n, lk, dv) -> (n, lq, dv)
+    return mx.sym.batch_dot(lhs=probs, rhs=values)
+
+
+class MultiHeadAttentionBase:
```

Comment:
no because also used for the non-self attention. Self attention is a special case of MultiHeadAttention where the queries == keys/values